### PR TITLE
unify authn for xds and sds clients

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pkg/security/authentication.go
Comment (hzxuzhonghu) [inline pkg/security/authentication.go:1]: this is originally from pilot/pkg/xds/auth.go

**Review by costinm — COMMENTED** [review]
Not sure XDS and SDS/CA clients should have the same behavior. 

For XDS - it is fine to use lighter auth, there is nothing secret - just configs and endpoints. 

For SDS/CA - we want more strict auth. Unfortunately we don't do a lot of checks - like 
verify the IP address ( now that we have the reverse index ), but should become stronger.

Not strongly against this change - IMO longer term it would be best to put a gateway in front of Istiod and use our own authn/authz instead of custom.

**Review by howardjohn — CHANGES_REQUESTED** [review]

File: security/pkg/nodeagent/caclient/providers/citadel/client_test.go [hunk 1]
```diff
@@ -65,10 +65,9 @@ type mockCAServer struct {
 
 func (ca *mockCAServer) CreateCertificate(ctx context.Context, in *pb.IstioCertificateRequest) (*pb.IstioCertificateResponse, error) {
 	if ca.Authenticator != nil {
-		am := security.AuthenticationManager{Authenticators: []security.Authenticator{ca.Authenticator}}
-		caller := am.Authenticate(ctx)
+		caller, err := security.Authenticate(ctx, []security.Authenticator{ca.Authenticator})
 		if caller == nil {
-			return nil, status.Error(codes.Unauthenticated, "request authenticate failure")
+			return nil, status.Error(codes.Unauthenticated, err.Error())
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply security/pkg/nodeagent/caclient/providers/citadel/client_test.go:70]: This is the test, i can restore it too

File: pkg/security/authentication.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	"errors"
+
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/peer"
+
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pkg/env"
+)
+
+var AuthPlaintext = env.Register("XDS_AUTH_PLAINTEXT", false,
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pkg/security/authentication.go:28]: It does not change sds. YOu can see the return value of Authenticate for CA server is nil, nil when it is through 15010(plaintext)

But we donot allow nil for SDS

File: security/pkg/server/ca/server.go [hunk 3]
```diff
@@ -77,17 +76,9 @@ func (s SaNode) String() string {
...
-	}
-
 	s.monitoring.CSR.Increment()
-	am := security.AuthenticationManager{Authenticators: s.Authenticators}
-	caller := am.Authenticate(ctx)
-	if caller == nil {
-		serverCaLog.Errorf("Failed to authenticate client from %s: %s", peerAddr, am.FailedMessages())
-		s.monitoring.AuthnError.Increment()
+	caller, err := security.Authenticate(ctx, s.Authenticators)
+	if caller == nil || err != nil {
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply security/pkg/server/ca/server.go:81]: @howardjohn It does not allow nil caller

**Comment by hzxuzhonghu** [comment]
@howardjohn ptaL

File: pkg/security/authentication.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	"authenticate plain text requests - used if Istiod is running on a secure/trusted network").Get()
+
+// Authenticate authenticates the ADS request using the configured authenticators.
+// Returns the validated principals or an error.
+// If no authenticators are configured, or if the request is on a non-secure
+// stream ( 15010 ) - returns an empty list of principals and no errors.
```
Comment (howardjohn) [inline pkg/security/authentication.go:34]: nit: "empty list of principals" should be updated, its now a nil struct

File: security/pkg/server/ca/server.go [hunk 5]
```diff
@@ -77,17 +76,9 @@ func (s SaNode) String() string {
...
-	if peerInfo, ok := peer.FromContext(ctx); ok {
-		peerAddr = peerInfo.Addr.String()
-	}
-
 	s.monitoring.CSR.Increment()
-	am := security.AuthenticationManager{Authenticators: s.Authenticators}
-	caller := am.Authenticate(ctx)
-	if caller == nil {
-		serverCaLog.Errorf("Failed to authenticate client from %s: %s", peerAddr, am.FailedMessages())
-		s.monitoring.AuthnError.Increment()
```
Comment (howardjohn) [inline security/pkg/server/ca/server.go:90]: seems we lost this?

**Review by howardjohn — APPROVED** [review]
LGTM aside from metric. Can you fix and remove hold?

File: security/pkg/server/ca/server.go (hunk 5 above)
Comment (hzxuzhonghu) [reply security/pkg/server/ca/server.go:90]: good catch

**Pull request merged**